### PR TITLE
#6184 Fix - Handle massive images in DeckPicker Background

### LINKED ISSUE #6184: [2.10.1] Handle massive images in DeckPicker Background
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/c6ce8864-f070-4fc9-86e2-370556aad97f

100MB image crashed the DeckPicker

```
ava.lang.RuntimeException: Canvas: trying to draw too large(115741696bytes) bitmap.
at android.view.DisplayListCanvas.throwIfCannotDraw(DisplayListCanvas.java:229)
at android.view.RecordingCanvas.drawBitmap(RecordingCanvas.java:101)
at android.graphics.drawable.BitmapDrawable.draw(BitmapDrawable.java:545)
at android.view.View.getDrawableRenderNode(View.java:22217)
at android.view.View.drawBackground(View.java:22123)
at android.view.View.draw(View.java:21879)
at android.view.View.updateDisplayListIfDirty(View.java:20761)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at androidx.drawerlayout.widget.DrawerLayout.drawChild(DrawerLayout.java:14)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.draw(View.java:21891)
at android.view.View.updateDisplayListIfDirty(View.java:20761)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.updateDisplayListIfDirty(View.java:20747)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.updateDisplayListIfDirty(View.java:20747)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.updateDisplayListIfDirty(View.java:20747)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.updateDisplayListIfDirty(View.java:20747)
at android.view.View.draw(View.java:21614)
at android.view.ViewGroup.drawChild(ViewGroup.java:4558)
at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4333)
at android.view.View.draw(View.java:21891)
at com.android.internal.policy.DecorView.draw(DecorView.java:1082)
at android.view.View.updateDisplayListIfDirty(View.java:20761)
at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:725)
at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:731)
at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:840)
at android.view.ViewRootImpl.draw(ViewRootImpl.java:3935)
at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3709)
at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3017)
at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1876)
at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8499)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:986)
at android.view.Choreographer.doCallbacks(Choreographer.java:764)
at android.view.Choreographer.doFrame(Choreographer.java:699)
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:965)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7073)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:494)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
```

### PR DESCRIPTION
**Pull request by infinyte7**
## Purpose / Description
Handle massive images in DeckPicker Background

## Fixes
Fixes #6184

## Approach
Reducing copied image size in AnkiDroid directory

## How Has This Been Tested?
Tested on moto x4
1. Ankdroid -> Setting -> Appearnce -> Background
2. Select an image with large size.
For ex:
```
https://visibleearth.nasa.gov/images/73751/july-blue-marble-next-generation-w-topography-and-bathymetry
```
3. Image get reduced in AnkiDroid folder.
``` For file size
284 MB
size reduced to 
321 KB
```
reduced

## Learning (optional, can help others)
https://android.jlelse.eu/loading-large-bitmaps-efficiently-in-android-66826cd4ad53
https://stackoverflow.com/questions/18573774/how-to-reduce-an-image-file-size-before-uploading-to-a-server

## Checklist
_Please, go through these checks before submitting the PR._

- [ x ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ x ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ x ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ x ] You have commented your code, particularly in hard-to-understand areas
- [ x ] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 1]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
         }
     }
 
+    private File compressedImage(File file) {
+        try {
+            BitmapFactory.Options bitmap = new BitmapFactory.Options();
+            bitmap.inJustDecodeBounds = true;
+            bitmap.inSampleSize = 6;
+
+            FileInputStream inputStream = new FileInputStream(file);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:462]: use try with resources

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 2]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+            BitmapFactory.decodeStream(inputStream, null, bitmap);
+            inputStream.close();
+
+            final int REQUIRED_SIZE=75;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:466]: Could we get a comment explaining the variable (what units?) Surely not pixels

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 3]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
                 UIUtils.showThemedToast(this, getString(R.string.no_image_selected), false);
             }
         } catch (OutOfMemoryError | Exception e) {
             UIUtils.showThemedToast(this, getString(R.string.error_selecting_image, e.getLocalizedMessage()), false);
         }
     }
 
+    private File compressedImage(File file) {
+        try {
+            BitmapFactory.Options bitmap = new BitmapFactory.Options();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:458]: This is a complex procedure, could you explain what it does, and why it does it?

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 4]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+
+            file.createNewFile();
+
+            FileOutputStream outputStream = new FileOutputStream(file);
+
+            selectedBitmap.compress(Bitmap.CompressFormat.JPEG, 100 , outputStream);
+
+            return file;
+        } catch (Exception e) {
+            UIUtils.showThemedToast(this, getString(R.string.error_selecting_image, e.getLocalizedMessage()), false);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:489]: A toast is showed twice, once here, and once when null is returned

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 5]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+
+            int scale = 1;
+            while(bitmap.outWidth / scale / 2 >= REQUIRED_SIZE &&
+                    bitmap.outHeight / scale / 2 >= REQUIRED_SIZE) {
+                scale *= 2;
+            }
+
+            BitmapFactory.Options bitmap2 = new BitmapFactory.Options();
+            bitmap2.inSampleSize = scale;
+            inputStream = new FileInputStream(file);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:476]: try with resources here as well

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 6]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+            inputStream.close();
+
+            file.createNewFile();
+
+            FileOutputStream outputStream = new FileOutputStream(file);
+
+            selectedBitmap.compress(Bitmap.CompressFormat.JPEG, 100 , outputStream);
+
+            return file;
+        } catch (Exception e) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:488]: Also catch `OutOfMemoryError` here

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 7]
```diff
@@ -428,21 +430,68 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
             } else {
+                backgroundImage.setChecked(false);
                 UIUtils.showThemedToast(this, getString(R.string.no_image_selected), false);
             }
         } catch (OutOfMemoryError | Exception e) {
             UIUtils.showThemedToast(this, getString(R.string.error_selecting_image, e.getLocalizedMessage()), false);
         }
     }
 
+    private File compressedImage(File file) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:456]: Can you generalize this function and put it in BitmapUtils, like this one? https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java#L40

Especially important to pay attention to the error handling there, anything opened should be closed in a finally, for instance - also, is this potentially crunching the image too much? I actually have the blue marble as my desktop background and if I had selected that but then it got JPEG-blocked out with an overzealous crunch that I couldn't configure, I'd be less happy with it personally

Comment (infinyte7) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:456]: Can we put limit on file size? ( For ex: only image size less than 10 mb allowed)
Using this.
```java
File file = new File(filepath);
long length = file.length();
```

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:456]: Is our issue the filesize, or the size of the bitmap in memory?

**Comment by mikehardy** [comment]
> The main issue: What about users that are currently experiencing this issue?
> 
> I have a few ideas, but please suggest your own for discussion, will look to @mikehardy for guidance

Check if it's over a certain size before loading it, if it is, run it through this compressor as if they were picking it for the first time, then continue

**Comment by infinyte7** [comment]
Okay, I will implement this.

**Comment by infinyte7** [comment]
There problem with compress function, it rotate images. I have done this.

**Comment by david-allison-1** [comment]
Build is failed - missing String resource

```
C:\Users\travis\build\ankidroid\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\Preferences.java:439: error: cannot find symbol
                        UIUtils.showThemedToast(this, getString(R.string.image_max_size_10_mb), false);
                                                                        ^
  symbol:   variable image_max_size_10_mb
  location: class string
```

**Comment by infinyte7** [comment]
I updated it.

File: AnkiDroid/src/main/res/values/03-dialogs.xml [hunk 8]
```diff
@@ -184,7 +184,8 @@
     <string name="error_selecting_image">Error selecting image. Please see manual. %s</string>
     <string name="error_deleting_image">Error deleting image</string>
     <string name="failed_to_apply_background_image">Failed to apply background image %s</string>
-
+    <string name="image_max_size_10_mb">Maximum image size 10 MB allowed</string>
```
Comment (david-allison-1) [inline AnkiDroid/src/main/res/values/03-dialogs.xml:187]: Supply a format string so the variable can change without updating translations

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java [hunk 9]
```diff
@@ -421,21 +421,30 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
                     String imgPathString = cursor.getString(columnIndex);
                     File sourceFile = new File(imgPathString);
 
+                    // file size in MB
+                    long fileLength = sourceFile.length() / (1024 * 1024);
+
                     String currentAnkiDroidDirectory = CollectionHelper.getCurrentAnkiDroidDirectory(this);
                     String imageName = "DeckPickerBackground.png";
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java:428]: consider moving this to a `private static final String` variable

**Comment by david-allison-1** [comment]
As a stopgap, I think we'd be best to gate the background image selection in the DeckPicker on a preference as well as the existence of the file, then a user can remove their data to fix the issue. It's not ideal, but it's better than the situation that they're in now.

**Comment by infinyte7** [comment]
Okay, But it will be beneficial to add restriction on file size. Because when homescreen load then it may be slow. And limiting on file size selection prevents adding large image file size. It will be better in loading decpicker activity relatively faster.

**Comment by david-allison-1** [comment]
Definitely! We should do both:

* We shouldn't let the user import a bad image (size limit on the input), and ideally when loading the image on the DeckPicker
* But, if they get around these restrictions, we want reinstalling the app to fix it.

**Comment by infinyte7** [comment]
yes, but deleting DeckPickerBackground.png  image from folder is also can fix.

**Comment by david-allison-1** [comment]
> yes, but deleting DeckPickerBackground.png image from folder is also can fix.

Users won't necessarily know how (or take the effort to find out how) to do this if the app is crashing, IMO, we should aim to make this as simple as possible for users to fix on their own.

**Comment by infinyte7** [comment]
Okay, I understand this.

**Comment by david-allison-1** [comment]
Okay, this has been hanging for a while. The code is better as it is now. I'll vote to merge it as-is and follow-up with a tactical PR to help any users who are currently crashing.

**Pull request merged**